### PR TITLE
Unify Timespan dialogs

### PR DESCRIPTION
**Pull request by RoiEXLab**
Follow-up of #2429
I originally wanted to create some tests for this, but I actually needed very few code.
The only thing that can be tested is perhaps the TimeUnit enum, I will add some tests for it later/tomorrow when I have some time again.
I just wanted to get some general Feedback on this change.

EDIT: This is what the dialog now looks like:
Screenshot1
Screenshot2

### CODE REVIEW

**Comment by RoiEXLab** [comment]
Ok I added some tests, this PR should now be ready for review...

**Comment by RoiEXLab** [comment]
Not as much as I hoped for, but I have no clue what else could be tested...

File: src/main/java/games/strategy/triplea/ui/TimespanDialog.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    }
+  }
+
+  /**
+   * Prompts the user to enter a timespan.
+   * If the operation is not cancelled, the action Consumer is run.
+   * Not that the Date passed to the consumer can be null if the user chose forever.
+   */
+  public void prompt(final Component parent, final String title, final String infoMessage,
+      final Consumer<Date> action) {
```
Comment (ssoloff) [inline src/main/java/games/strategy/triplea/ui/TimespanDialog.java]: There are a few more indications here that this dialog is lobby ban/mute-specific:

* It uses `Date` as the type of the ban timestamp.
* It uses `null` to indicate forever.

Thus, it seems like this type should be renamed to indicate it is specific to lobby ban/mute timestamps or (probably better) moved to the `g.s.engine.lobby.client.ui` package.

File: src/main/java/games/strategy/triplea/ui/TimespanDialog.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    DAYS("Days", i -> Instant.now().plus(i, ChronoUnit.DAYS)),
+    WEEKS("Weeks", i -> LocalDateTime.now(ZoneOffset.UTC).plus(i, ChronoUnit.WEEKS).toInstant(ZoneOffset.UTC)),
+    MONTHS("Months", i -> LocalDateTime.now(ZoneOffset.UTC).plus(i, ChronoUnit.MONTHS).toInstant(ZoneOffset.UTC)),
+    YEARS("Years", i -> LocalDateTime.now(ZoneOffset.UTC).plus(i, ChronoUnit.YEARS).toInstant(ZoneOffset.UTC)),
+    FOREVER("Forever", i -> null);
+
+    private final String name;
+    private final Function<Integer, Instant> function;
+
+    private TimeUnit(final String name, final Function<Integer, Instant> supplier) {
```
Comment (ssoloff) [inline src/main/java/games/strategy/triplea/ui/TimespanDialog.java]: Should the parameter `supplier` be named `function` here?

File: src/test/java/games/strategy/engine/lobby/client/ui/LobbyFrameTest.java [hunk 3]
```diff
@@ -1,91 +0,0 @@
...
-import static org.hamcrest.Matchers.is;
-import static org.junit.Assert.assertThat;
-
-import java.time.Instant;
-
-import org.junit.Test;
-
-import games.strategy.engine.lobby.client.ui.LobbyFrame.TimeUnitNames;
-
-public final class LobbyFrameTest {
```
Comment (ssoloff) [inline src/test/java/games/strategy/engine/lobby/client/ui/LobbyFrameTest.java]: A little background on these tests...  They were added to cover the very subtle `UnsupportedTemporalTypeException` that can occur at runtime when using non-constant periodic temporal types like `ChronoUnit#MONTHS`.  That's why we have to use `LocalDateTime` as an intermediate type for those types of units.  These tests were there to prevent regressions in the event someone came along and changed the implementation.  It seems like that could still be possible with the `TimespanDialog$TimeUnit` implementations.

File: src/test/java/games/strategy/triplea/ui/TimespanDialogTest.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  }
+
+  @Test
+  public void testIntervalSizesBecomeBigger() {
+    final Instant minute = TimeUnit.MINUTES.getInstant(1);
+    final Instant hour = TimeUnit.HOURS.getInstant(1);
+    final Instant day = TimeUnit.DAYS.getInstant(1);
+    final Instant week = TimeUnit.WEEKS.getInstant(1);
+    final Instant month = TimeUnit.MONTHS.getInstant(1);
+    final Instant year = TimeUnit.YEARS.getInstant(1);
```
Comment (ssoloff) [inline src/test/java/games/strategy/triplea/ui/TimespanDialogTest.java]: Note that all these instants are potentially being created with a different start time since the call to `Instant#now()` is buried inside the `getInstant()` calls.  As we discussed before, tests shouldn't be dependent on wall time.  It might be better for `TimeUnit#getInstant()` to take the start time (in addition to the duration) and require the caller to pass it in.  That will push the `Instant#now()` calls into a single location within `TimespanDialog#prompt()`.  Something like this:

```diff
diff --git a/src/main/java/games/strategy/triplea/ui/TimespanDialog.java b/src/main/java/games/strategy/triplea/ui/TimespanDialog.java
index 92cfc30..3b6d97f 100644
--- a/src/main/java/games/strategy/triplea/ui/TimespanDialog.java
+++ b/src/main/java/games/strategy/triplea/ui/TimespanDialog.java
@@ -6,6 +6,7 @@ import java.time.LocalDateTime;
 import java.time.ZoneOffset;
 import java.time.temporal.ChronoUnit;
 import java.util.Date;
+import java.util.function.BiFunction;
 import java.util.function.Consumer;
 import java.util.function.Function;
 
@@ -30,18 +31,18 @@ public class TimespanDialog {
    */
   @VisibleForTesting
   enum TimeUnit {
-    MINUTES("Minutes", i -> Instant.now().plus(i, ChronoUnit.MINUTES)),
-    HOURS("Hours", i -> Instant.now().plus(i, ChronoUnit.HOURS)),
-    DAYS("Days", i -> Instant.now().plus(i, ChronoUnit.DAYS)),
-    WEEKS("Weeks", i -> LocalDateTime.now(ZoneOffset.UTC).plus(i, ChronoUnit.WEEKS).toInstant(ZoneOffset.UTC)),
-    MONTHS("Months", i -> LocalDateTime.now(ZoneOffset.UTC).plus(i, ChronoUnit.MONTHS).toInstant(ZoneOffset.UTC)),
-    YEARS("Years", i -> LocalDateTime.now(ZoneOffset.UTC).plus(i, ChronoUnit.YEARS).toInstant(ZoneOffset.UTC)),
-    FOREVER("Forever", i -> null);
+    MINUTES("Minutes", (start, i) -> start.plus(i, ChronoUnit.MINUTES)),
+    HOURS("Hours", (start, i) -> start.plus(i, ChronoUnit.HOURS)),
+    DAYS("Days", (start, i) -> start.plus(i, ChronoUnit.DAYS)),
+    WEEKS("Weeks", (start, i) -> LocalDateTime.ofInstant(start, ZoneOffset.UTC).plus(i, ChronoUnit.WEEKS).toInstant(ZoneOffset.UTC)),
+    MONTHS("Months", (start, i) -> LocalDateTime.ofInstant(start, ZoneOffset.UTC).plus(i, ChronoUnit.MONTHS).toInstant(ZoneOffset.UTC)),
+    YEARS("Years", (start, i) -> LocalDateTime.ofInstant(start, ZoneOffset.UTC).plus(i, ChronoUnit.YEARS).toInstant(ZoneOffset.UTC)),
+    FOREVER("Forever", (start, i) -> null);
 
     private final String name;
-    private final Function<Integer, Instant> function;
+    private final BiFunction<Instant, Integer, Instant> function;
 
-    private TimeUnit(final String name, final Function<Integer, Instant> supplier) {
+    private TimeUnit(final String name, final BiFunction<Instant, Integer, Instant> supplier) {
       this.name = name;
       this.function = supplier;
     }
@@ -52,8 +53,8 @@ public class TimespanDialog {
     }
 
     @VisibleForTesting
-    Instant getInstant(final Integer integer) {
-      return function.apply(integer);
+    Instant getInstant(final Instant start, final Integer integer) {
+      return function.apply(start, integer);
     }
   }
 
@@ -79,7 +80,7 @@ public class TimespanDialog {
             .build())
         .build(), title, JOptionPane.OK_CANCEL_OPTION);
     if (returnType == JOptionPane.OK_OPTION) {
-      final Instant instant = ((TimeUnit) comboBox.getSelectedItem()).getInstant((Integer) spinner.getValue());
+      final Instant instant = ((TimeUnit) comboBox.getSelectedItem()).getInstant(Instant.now(), (Integer) spinner.getValue());
       action.accept(instant == null ? null : Date.from(instant));
     }
   }
```

File: src/main/java/games/strategy/triplea/ui/TimespanDialog.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+      this.name = name;
+      this.function = supplier;
+    }
+
+    @Override
+    public String toString() {
+      return name;
+    }
+
+    @VisibleForTesting
```
Comment (ssoloff) [inline src/main/java/games/strategy/triplea/ui/TimespanDialog.java]: Is this annotation needed?  This method is called from `prompt()`, so it's not really visible just for test purposes.  (Obviously, it could be called from `prompt()` even if it were marked private, but that smells a bit as the compiler will still generate a synthetic package-private pass-through method so the private method can be called at runtime.)

File: src/main/java/games/strategy/triplea/ui/TimespanDialog.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+   * Prompts the user to enter a timespan.
+   * If the operation is not cancelled, the action Consumer is run.
+   * Not that the Date passed to the consumer can be null if the user chose forever.
+   */
+  public void prompt(final Component parent, final String title, final String infoMessage,
+      final Consumer<Date> action) {
+    final JSpinner spinner = new JSpinner(new SpinnerNumberModel(0, 0, Integer.MAX_VALUE, 1));
+    final JComboBox<TimeUnit> comboBox = new JComboBox<>(TimeUnit.values());
+    comboBox.addActionListener(e -> spinner.setEnabled(!comboBox.getSelectedItem().equals(TimeUnit.FOREVER)));
+    final int returnType = JOptionPane.showConfirmDialog(parent, JPanelBuilder.builder()
```
Comment (ssoloff) [inline src/main/java/games/strategy/triplea/ui/TimespanDialog.java]: Not sure return "type" makes sense in the variable name here.  Would `returnValue` or `selectedOption` be better?

File: src/test/java/games/strategy/triplea/ui/TimespanDialogTest.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+import org.junit.Test;
+
+import games.strategy.triplea.ui.TimespanDialog.TimeUnit;
+
+public class TimespanDialogTest {
+
+  @Test
+  public void testForeverReturnsNull() {
+    assertNull(TimeUnit.FOREVER.getInstant((int) (Math.random() * Integer.MAX_VALUE)));
```
Comment (ssoloff) [inline src/test/java/games/strategy/triplea/ui/TimespanDialogTest.java]: Random inputs tend to be a smell in unit tests.  I would just test the boundary conditions (`Integer#MIN_VALUE`, `Integer#MAX_VALUE`) and zero if you're trying to demonstrate that this method will always return `null`.

File: src/test/java/games/strategy/triplea/ui/TimespanDialogTest.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    assertTrue(day.isBefore(week));
+    assertTrue(week.isBefore(month));
+    assertTrue(month.isBefore(year));
+  }
+
+  @Test
+  public void testPositiveIntIsInFuture() {
+    assertTrue(Instant.now().isBefore(TimeUnit.MINUTES.getInstant(1)));
+  }
+}
```
Comment (ssoloff) [inline src/test/java/games/strategy/triplea/ui/TimespanDialogTest.java]: If you want to get some additional test coverage out of the logic in `prompt()`, you could extract the last few lines of that method so you could test the interaction with the `Consumer`:

* Verify consumer is not called when dialog is cancelled.
* Verify consumer receives `null` when expiration time is `null`.
* Verify consumer receives the correct `Date` (or just a non-`null` value) when expiration time is non-`null`.

The extraction would look something like:

```java
diff --git a/src/main/java/games/strategy/triplea/ui/TimespanDialog.java b/src/main/java/games/strategy/triplea/ui/TimespanDialog.java
index 92cfc30..d17d10f 100644
--- a/src/main/java/games/strategy/triplea/ui/TimespanDialog.java
+++ b/src/main/java/games/strategy/triplea/ui/TimespanDialog.java
@@ -78,8 +78,13 @@ public class TimespanDialog {
             .add(comboBox)
             .build())
         .build(), title, JOptionPane.OK_CANCEL_OPTION);
+    runAction(action, returnType, (TimeUnit) comboBox.getSelectedItem(), (Integer) spinner.getValue());
+  }
+
+  @VisibleForTesting
+  void runAction(final Consumer<Date> action, final int returnType, final TimeUnit timeUnit, final Integer duration) {
     if (returnType == JOptionPane.OK_OPTION) {
-      final Instant instant = ((TimeUnit) comboBox.getSelectedItem()).getInstant((Integer) spinner.getValue());
+      final Instant instant = timeUnit.getInstant(duration);
       action.accept(instant == null ? null : Date.from(instant));
     }
   }
```

File: src/main/java/games/strategy/triplea/ui/TimespanDialog.java (hunk 5 above)
Comment (RoiEXLab) [reply src/main/java/games/strategy/triplea/ui/TimespanDialog.java]: From the javadocs of `@VisibleForTesting`:
> Annotates a program element that exists, or is more widely visible than otherwise necessary, only for use in test code.

If I mark this method as private, I can still call it from TimespanDialog, but no other classes, I just raised the visibility to be able to call this methods from tests.
So in my opinion this annotation is 100% appropriate

**Comment by RoiEXLab** [comment]
@ssoloff Updated!

File: src/main/java/games/strategy/triplea/ui/TimespanDialog.java (hunk 5 above)
Comment (ssoloff) [reply src/main/java/games/strategy/triplea/ui/TimespanDialog.java]: I agree, you _can_ call it from `TimespanDialog` if it's private.  However, this is considered bad practice in some cases because the compiler actually generates a second package-private method (a so-called "synthetic accessor") that simply passes through to the private method.  There's even a warning you can turn on in most static analyzers for that (in Eclipse, it's under **Code Style > Access to a non-accessible member of an enclosing type**).

NBD though.  It just caught my attention because I expected a private method to be called by the class that declared it.  Otherwise, if it's being called by other types (including enclosing types), then that seems to imply the method is part of the class API.

File: src/main/java/games/strategy/engine/lobby/client/ui/TimespanDialog.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+   * Prompts the user to enter a timespan.
+   * If the operation is not cancelled, the action Consumer is run.
+   * Not that the Date passed to the consumer can be null if the user chose forever.
+   */
+  public static void prompt(final Component parent, final String title, final String infoMessage,
+      final Consumer<Date> action) {
+    final JSpinner spinner = new JSpinner(new SpinnerNumberModel(0, 0, Integer.MAX_VALUE, 1));
+    final JComboBox<TimeUnit> comboBox = new JComboBox<>(TimeUnit.values());
+    comboBox.addActionListener(e -> spinner.setEnabled(!comboBox.getSelectedItem().equals(TimeUnit.FOREVER)));
+    final int returnValue = JOptionPane.showConfirmDialog(parent, JPanelBuilder.builder()
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/lobby/client/ui/TimespanDialog.java]: A nice UX improvement here would be to give focus to the spinner by default (right now it goes to the OK button).  This seems to be more difficult than it needs to be when using `JOptionPane` (e.g. see [here](https://stackoverflow.com/questions/6251665/setting-component-focus-in-joptionpane-showoptiondialog); and some of those solutions don't seem to work on Linux).

It may be required to create a custom `JDialog` component to get that to work (e.g. similar to what we did for `ProgressDialog`) instead of piggybacking off of `JOptionPane`.

File: src/test/java/games/strategy/engine/lobby/client/ui/TimespanDialogTest.java [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+        JOptionPane.OK_OPTION, TimeUnit.MINUTES, Integer.MAX_VALUE);
+    TimespanDialog.runAction(d -> assertTrue(d.after(new Date())),
+        JOptionPane.OK_OPTION, TimeUnit.HOURS, Integer.MAX_VALUE);
+    TimespanDialog.runAction(d -> assertTrue(d.after(new Date())),
+        JOptionPane.OK_OPTION, TimeUnit.DAYS, Integer.MAX_VALUE);
+    TimespanDialog.runAction(d -> assertTrue(d.after(new Date())),
+        JOptionPane.OK_OPTION, TimeUnit.WEEKS, Integer.MAX_VALUE);
+    TimespanDialog.runAction(d -> assertTrue(d.after(new Date())),
+        JOptionPane.OK_OPTION, TimeUnit.MONTHS, Integer.MAX_VALUE);
+    // We can use Integer#MAX_VALUE for years, because this will result in a long-overflow
```
[the remark this answers is not in the document]
Comment (RoiEXLab) [reply src/test/java/games/strategy/engine/lobby/client/ui/TimespanDialogTest.java]: Yes, typos 🙈

**Comment by ssoloff** [comment]
I tried to ban a username for 2,147,483,647 years and received the following exception:

```
java.time.DateTimeException: Invalid value for Year (valid values -999999999 - 999999999): 2147485664
	at java.time.temporal.ValueRange.checkValidIntValue(ValueRange.java:330)
	at java.time.temporal.ChronoField.checkValidIntValue(ChronoField.java:722)
	at java.time.LocalDate.plusYears(LocalDate.java:1277)
	at java.time.LocalDate.plus(LocalDate.java:1241)
	at java.time.LocalDateTime.plus(LocalDateTime.java:1194)
	at games.strategy.engine.lobby.client.ui.TimespanDialog$TimeUnit.lambda$5(TimespanDialog.java:38)
	at games.strategy.engine.lobby.client.ui.TimespanDialog$TimeUnit.getInstant(TimespanDialog.java:56)
	at games.strategy.engine.lobby.client.ui.TimespanDialog.runAction(TimespanDialog.java:88)
	at games.strategy.engine.lobby.client.ui.TimespanDialog.prompt(TimespanDialog.java:81)
	at games.strategy.triplea.ui.menubar.LobbyMenu.lambda$1(LobbyMenu.java:162)
	at javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:2022)
        < . . . snip . . . >
```

We might have to support a custom range of allowable values per `TimeUnit`.  If so, unit tests would be good here to test the boundary conditions.

File: src/test/java/games/strategy/engine/lobby/client/ui/TimespanDialogTest.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (prastle) [reply src/test/java/games/strategy/engine/lobby/client/ui/TimespanDialogTest.java]: 2 billion plus years ;) Sorry had to laugh.  👍

**Comment by RoiEXLab** [comment]
@ssoloff I implemented a modified version of the ancestorlistener,  it works for me (Windows 10 64 bit).
Also I simply decreased the int limit for all timeunits. 99999999 seems to work for all timeunits without prroblems, and if a ban should take a long time, the next-higher timeunit can be chosen

File: src/main/java/games/strategy/engine/lobby/client/ui/TimespanDialog.java [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+
+  /**
+   * Prompts the user to enter a timespan.
+   * If the operation is not cancelled, the action Consumer is run.
+   * Not that the Date passed to the consumer can be null if the user chose forever.
+   */
+  public static void prompt(final Component parent, final String title, final String infoMessage,
+      final Consumer<Date> action) {
+    final JSpinner spinner = new JSpinner(new SpinnerNumberModel(0, 0, 99999999, 1));
+    spinner.addAncestorListener(new AncestorListener() {
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/lobby/client/ui/TimespanDialog.java]: Unfortunately, this does not work on Linux. 😞  There's apparently a known wont-fix [Java bug](http://bugs.java.com/bugdatabase/view_bug.do?bug_id=5018574) related to this.

File: src/main/java/games/strategy/engine/lobby/client/ui/TimespanDialog.java [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+
+      @Override
+      public void ancestorAdded(AncestorEvent e) {
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/lobby/client/ui/TimespanDialog.java]: A few method parameters in this class can be `final`.

File: src/main/java/games/strategy/engine/lobby/client/ui/TimespanDialog.java [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+  }
+
+  /**
+   * Prompts the user to enter a timespan.
+   * If the operation is not cancelled, the action Consumer is run.
+   * Not that the Date passed to the consumer can be null if the user chose forever.
+   */
+  public static void prompt(final Component parent, final String title, final String infoMessage,
+      final Consumer<Date> action) {
+    final JSpinner spinner = new JSpinner(new SpinnerNumberModel(0, 0, 99999999, 1));
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/lobby/client/ui/TimespanDialog.java]: Might be good to extract `99999999` to a `@VisibleForTesting` constant so it isn't duplicated throughout the test code (and maybe write it as `99_999_999` so its magnitude is obvious at a glance without having to count the number of digits :smile:).

File: src/main/java/games/strategy/engine/lobby/client/ui/TimespanDialog.java (hunk 11 above)
Comment (RoiEXLab) [reply src/main/java/games/strategy/engine/lobby/client/ui/TimespanDialog.java]: Could you test it with `SwingUtilities.invokeLater(e.getComponent()::requestFocus);` ?
(Actually not sure this is correct syntax)

Comment (ssoloff) [reply src/main/java/games/strategy/engine/lobby/client/ui/TimespanDialog.java]: Did that. :smile:  I tested with both workarounds given in the bug report, including several permutations with `invokeLater()` around various blocks of code.  No joy.

Comment (ssoloff) [reply src/main/java/games/strategy/engine/lobby/client/ui/TimespanDialog.java]: And, yes, I tried both `requestFocus()` and `requestFocusInWindow()`. :smile:

Comment (RoiEXLab) [reply src/main/java/games/strategy/engine/lobby/client/ui/TimespanDialog.java]: I really don't want to create my own custom Dialog.
Having a simple blocking method that returns the "exit" code is so handy.
If this bothers you too much, you can of course submit a PR that adds a general purpose Utility class instead of JOptionPane which offers a similar behaviour, but allows to set the default selected component

Comment (ssoloff) [reply src/main/java/games/strategy/engine/lobby/client/ui/TimespanDialog.java]: Agreed, it's not a big deal right now.  I'll add it to my Wish List. :smile:

File: src/main/java/games/strategy/engine/lobby/client/ui/LobbyFrame.java [hunk 14]
```diff
@@ -134,124 +119,55 @@ private void showServerMessage(final LobbyServerProperties props) {
...
-      final Instant expire = addDuration(Instant.now(), result2, selectedTimeUnit);
-      if (selectedMuteType.toLowerCase().contains("name")) {
-        controller.muteUsername(clickedOn, Date.from(expire));
-      }
-      if (selectedMuteType.toLowerCase().contains("mac")) {
-        controller.muteMac(clickedOn, Date.from(expire));
-      }
+      TimespanDialog.prompt(this, "Select Timespan",
+          "Please consult other admins before muting longer than 1 day. \n"
+              + "And please remember to report this ban.",
```
[the remark this answers is not in the document]
Comment (ssoloff) [reply src/main/java/games/strategy/engine/lobby/client/ui/LobbyFrame.java]: Bump...

[the remark this answers is not in the document]
Comment (RoiEXLab) [reply src/main/java/games/strategy/engine/lobby/client/ui/LobbyFrame.java]: Overlooked the second typo ^^

File: src/main/java/games/strategy/engine/lobby/client/ui/TimespanDialog.java [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+    private final String name;
+    private final Function<Integer, Instant> function;
+
+    private TimeUnit(final String name, final Function<Integer, Instant> function) {
+      this.name = name;
+      this.function = function;
+    }
+
+    @Override
+    public String toString() {
```
Comment (DanVanAtta) [inline src/main/java/games/strategy/engine/lobby/client/ui/TimespanDialog.java]: Minor, I'm a bit confused why a `TimespanDialog` has a `toString` function,
Trying to answer that, I notice that the `name` value is passed in via constructor and then we simply return that. For some reason then this dialog is also a 'carrier' for this name value and is independent of everyting else in this class (unless I missed a usage somewhere which is very possible).

I'll try to come back to this PR tomorrow to give a more detailed feedback, apologies for not quite fully tracing out things to answer my own questions.

Comment (DanVanAtta) [reply src/main/java/games/strategy/engine/lobby/client/ui/TimespanDialog.java]: Ah, it's the toString() of the enum type... that helps.
At this point I would suggest we rename the `name` variable, it's not very useful for knowing what we are naming : )

File: src/main/java/games/strategy/engine/lobby/client/ui/TimespanDialog.java (hunk 13 above)
Comment (ssoloff) [inline src/main/java/games/strategy/engine/lobby/client/ui/TimespanDialog.java]: @RoiEXLab, forgot to bump this yesterday...  Just curious if setting the default value of the spinner to 1 instead of 0 would be better UX?  The old code used to default the duration to 1.  If the user just hits <kbd>Enter</kbd> with 0 set, no ban/mute will be occur.  Maybe that was your intention?

Comment (RoiEXLab) [reply src/main/java/games/strategy/engine/lobby/client/ui/TimespanDialog.java]: Idea was to have 0 as some sort of unban method, but there is a seperate menu option for this, so not really needed ^^

File: src/main/java/games/strategy/engine/lobby/client/ui/TimespanDialog.java (hunk 15 above)
Comment (RoiEXLab) [reply src/main/java/games/strategy/engine/lobby/client/ui/TimespanDialog.java]: The  toString method is veing invoked by the default cellrenderer of JComboBox.
Either the toString methods gets overriden (simpler), or we create our own CellRenderer to rendem our enum customly, in order to have a normal case (no caps)

Comment (DanVanAtta) [reply src/main/java/games/strategy/engine/lobby/client/ui/TimespanDialog.java]: I see, thanks.
I wonder why we are allowing just this much flexibility. Who really needs to specify number of years when you can just say 365 days? I would recommend long term we simplify this and create a code enforced ban policy.

First ban = 15 minutes, second = 1 day, third = 1 week, fourth = permanent.

Then mods will just click ban and the code will do the right thing. Along the way we also record these bans so we don't have the overhead of mods to report their actions.

Given all that, for what we have, the enum is actually pretty slick/reasonable, but I think the premise of what we are doing is a bit misguided with too much flexibility allowed (and hence all these warnings about asking other mods before doing more than 1 day bans, and also unclear how we can have a consistent ban policy if we are arbitrarily picking hours or days)

**Comment by DanVanAtta** [comment]
Sry, won't have time to finish reviewing this further than I have @RoiEXLab until Friday.  Given the one approval, not sure if we really need the extra eyes. Some play testing could help inform if we like all the updates and we can easily tweak it going forward.
